chain/chain/src/flat_storage_resharder.rs

```diff
@@ -2009,4 +2009,116 @@ mod tests {
     }
+
+    /// The split of a parent shard shouldn't happen until the resharding block has become final.
+    #[test]
+    #[ignore]
+    fn shard_split_should_wait_final_block() {
+        init_test_logger();
+        let (mut chain, resharder, sender) =
+            create_chain_resharder_sender::<DelayedSender>(simple_shard_layout());
+        let new_shard_layout = shard_layout_after_split();
+        let flat_store = resharder.runtime.store().flat_store();
+
+        // Add two blocks to the chain.
+        add_blocks_to_chain(&mut chain, 2);
+        assert_eq!(chain.head().unwrap().height, 2);
+        assert_eq!(chain.final_head().unwrap().height, 0);
+
+        // Trigger resharding at block 2 and it shouldn't split the parent shard.
+        let resharding_event_type = event_type_from_chain_and_layout(&chain, &new_shard_layout);
+        let ReshardingSplitShardParams { parent_shard, .. } = match resharding_event_type.clone() {
+            ReshardingEventType::SplitShard(params) => params,
+        };
+        assert!(resharder.start_resharding(resharding_event_type, &new_shard_layout).is_ok());
+        assert_eq!(sender.call_split_shard_task(), FlatStorageReshardingTaskStatus::Postponed);
+        assert!(flat_store.iter(parent_shard).count() > 0);
+
+        // Move final head to the resharding block (2) by adding more blocks.
+        add_blocks_to_chain(&mut chain, 2);
+        assert_eq!(chain.final_head().unwrap().height, 2);
+
+        // Trigger resharding again and now it should split the parent shard.
+        assert_eq!(
+            sender.call_split_shard_task(),
+            FlatStorageReshardingTaskStatus::Successful { num_batches_done: 1 }
+        );
+        assert_eq!(flat_store.iter(parent_shard).count(), 0);
+    }
+
+    /// Utility to add blocks on top of a chain.
+    fn add_blocks_to_chain(chain: &mut Chain, num_blocks: u64) {
+        let signer = Arc::new(create_test_signer("aa"));
+        let next_block_height = chain.head().unwrap().height + 1;
+        for height in next_block_height..next_block_height + num_blocks {
+            let prev_block = chain.get_block_by_height(height - 1).unwrap();
+            let block = TestBlockBuilder::new(Clock::real(), &prev_block, signer.clone())
+                .height(height)
+                .build();
+            chain.process_block_test(&None, block).unwrap();
+        }
+    }
+
+    /// Test to verify that a resharding event not yet started can be replaced by a newer resharding
+    /// event on a different resharding hash. This property is useful to have in the presence of
+    /// chain forks. For instance, the chain may wants to split a shard at some block B; there's a
+    /// chance B never becomes final and instead a new split is triggered at block B'. The latter
+    /// shouldn't be blocked by the presence of an earlier resharding event.
+    #[test]
+    #[ignore]
+    fn resharding_event_not_started_can_be_replaced() {
```

Comment:
That's awesome! Thinking about it, we should have unit tests like this one for each component and then one testloop integration test to exercise all in one place too. It's a great start here!